src/caffe/layers/data_layer.cpp

```diff
@@ -112,11 +118,12 @@ void DataLayer<Dtype>::load_batch(Batch<Dtype>* batch) {
   for (int item_id = 0; item_id < batch_size; ++item_id) {
       timer.Start();
       // Get a datum
-      Datum* datum = (reader_.full().pop("Waiting for data"));
+      string* str = (reader_.full().pop("Waiting for data"));
```

Comment:
That's handled here: https://github.com/slayton58/caffe/blob/thread_pool_parsing/src/caffe/data_reader.cpp#L112-L116

If you try to run past the end of the DB, you loop around and start again from the beginning